adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m

```diff
@@ -13,36 +13,75 @@
 // limitations under the License.
 
 #import "GADMediationAdapterPangle.h"
-
 #import "GADMediationAdapterPangleConstants.h"
+#import <BUAdSDK/BUAdSDK.h>
+#import "GADPangleRTBBannerRenderer.h"
+#import "GADPangleRTBInterstitialRenderer.h"
+#import "GADPangleRTBRewardedRenderer.h"
+#import "GADMediationAdapterPangleConstants.h"
+#import "GADMAdapterPangleUtils.h"
 
-@implementation GADMediationAdapterPangle
+@implementation GADMediationAdapterPangle {
+    /// pangle Audience Network banner ad wrapper.
+    GADPangleRTBBannerRenderer *_bannerRenderer;
+    /// pangle Audience Network interstitial ad wrapper.
+    GADPangleRTBInterstitialRenderer *_interstitialRenderer;
+    /// pangle Audience Network rewarded ad wrapper.
+    GADPangleRTBRewardedRenderer *_rewardedlRenderer;
+}
 
 - (void)collectSignalsForRequestParameters:(nonnull GADRTBRequestParameters *)params
                          completionHandler:(nonnull GADRTBSignalCompletionHandler)completionHandler {
-  // TODO: Start signal generation and call completionHandler when the signal generation finishes or fails.
-  completionHandler("", nil);
+    NSString *signals = [BUAdSDKManager getBiddingToken:nil];
+    if (completionHandler) {
+        completionHandler(signals,nil);
+    }
 }
 
 + (void)setUpWithConfiguration:(nonnull GADMediationServerConfiguration *)configuration
              completionHandler:(nonnull GADMediationAdapterSetUpCompletionBlock)completionHandler {
-  // TODO: Set up Pangle ad network SDK and perform any necessary prefetching or configuration work.
-  completionHandler(nil);
+    NSMutableSet *appIds = [[NSMutableSet alloc]init];
+    for (GADMediationCredentials *cred in configuration.credentials) {
+        NSString *appId = cred.settings[GADMAdapterPangleAppID];
+        if (appId && [appId isKindOfClass:[NSString class]] && appId.length) {
+            [appIds addObject:appId];
+        }
+    }
+    
+    if (appIds.count < 1) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorInvalidServerParameters, @"Pangle mediation configurations did not contain a valid App ID.");
+        completionHandler(error);
+        return;
+    }
+    
+    NSString *appID = [appIds anyObject];
+    if (appIds.count > 1) {
+        PangleLog(@"Found the following App IDs:%@. Please remove any app IDs which you are not using", appIds);
+        PangleLog(@"Configuring Pangle SDK with the app ID:%@", appID);
+    }
+    
+    BUAdSDKConfiguration *cog = [BUAdSDKConfiguration configuration];
+    cog.territory = BUAdSDKTerritory_NO_CN;
+    cog.appID = appID;
+    [BUAdSDKManager startWithAsyncCompletionHandler:^(BOOL success, NSError *error) {
+        completionHandler(error);
+    }];
```

Comment:
As per Pangle's setup document, we need to include any GDPR/COPPA/CCPA settings when initializing the SDK:

```
// COPPA
configuration.coppa = @(0); //0: child 1:adult
 
// GDPR
configuration.gdpr = @(0); //0: have access 1:no access

// CCPA
configuration.ccpa = @(0); //0: have access 1:no access
```

Does Pangle provide a way for publishers to set these privacy settings before/after initializing their SDK?

If not, then we need to expose a wrapper API to enable publishers to set these privacy settings prior to initializing the Pangle SDK.

See the AdColony mediation guide on how we did this for Android:
https://developers.google.com/admob/ios/mediation/adcolony#eu_consent_and_gdpr

Also, do we need additional code in the adapter to pass those settings to the SDK?

Reply:
Gentle bump on this, and see the below related discussion in the Android PR: 
https://github.com/googleads/googleads-mobile-android-mediation/pull/391/files#r817305797

One way to do it would be expose a wrapper API to enable publishers to set these privacy settings prior to initializing the Pangle SDK. 
See the AdColony mediation guide on how we did this for iOS:
https://developers.google.com/admob/ios/mediation/adcolony#eu_consent_and_gdpr